File: apps/mobile/src/lib/onboarding/logic/step-one-defaults.ts
/**
 * Mobile adaptation of web `lib/onboarding/step-one-defaults.ts`. The web's
 * sessionStorage locale-draft dance (for the mid-wizard locale switch that
 * remounts the app) is dropped — mobile keeps in-memory wizard state and does
 * not live-switch the app locale, so there's no remount to survive.
 */
import type { AppLocale } from '~/i18n';

const LOCALES: readonly AppLocale[] = ['en', 'vi'];
const DEFAULT_LOCALE: AppLocale = 'en';

export interface StepOneData {
  countryOfOrigin: string | null;
  countryOfResidence: string | null;
  preferredLocale: AppLocale;
}

function toSupportedLocale(value?: string | null): AppLocale | null {
  if (!value || !LOCALES.includes(value as AppLocale)) return null;
  return value as AppLocale;
}

export function buildStepOneDefaults({
  activeLocale,
  countryOfOrigin,
  countryOfResidence,
  profilePreferredLocale,
}: {
  activeLocale?: string | null;
  countryOfOrigin?: string | null;
  countryOfResidence?: string | null;
  profilePreferredLocale?: string | null;
}): StepOneData {
  return {
    countryOfOrigin: countryOfOrigin ?? null,
    countryOfResidence: countryOfResidence ?? null,
    preferredLocale:
      toSupportedLocale(activeLocale) ??
      toSupportedLocale(profilePreferredLocale) ??
      DEFAULT_LOCALE,
  };
}

File: apps/mobile-flutter/lib/features/onboarding/logic/step_one_defaults.dart
/// Mobile adaptation of RN `lib/onboarding/logic/step-one-defaults.ts`.
///
/// The web's sessionStorage locale-draft dance (for the mid-wizard locale switch
/// that remounts the app) is dropped — Flutter keeps in-memory wizard state and
/// does not live-switch the app locale, so there's no remount to survive.
library;

const List<String> _locales = ['en', 'vi'];
const String _defaultLocale = 'en';

class StepOneData {
  final String? countryOfOrigin;
  final String? countryOfResidence;
  final String preferredLocale;

  const StepOneData({
    required this.countryOfOrigin,
    required this.countryOfResidence,
    required this.preferredLocale,
  });
}

String? _toSupportedLocale(String? value) {
  if (value == null || !_locales.contains(value)) return null;
  return value;
}

StepOneData buildStepOneDefaults({
  String? activeLocale,
  String? countryOfOrigin,
  String? countryOfResidence,
  String? profilePreferredLocale,
}) {
  return StepOneData(
    countryOfOrigin: countryOfOrigin,
    countryOfResidence: countryOfResidence,
    preferredLocale: _toSupportedLocale(activeLocale) ??
        _toSupportedLocale(profilePreferredLocale) ??
        _defaultLocale,
  );
}
